# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек¶

Выполнил студент гр. 1304 Завражин Дмитрий. Вариант №34.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

### Понятие аддитивной цепочки

Аддитивной цепочкой натурального числа $n$ называется последовательность натуральных чисел
$$ 1=a_0,a_1,a_2,\dots,a_r=n,\quad   a_i=a_j+a_k, \quad \forall k \leq j < i$$

Длина цепочки обозначается как $r$.

Длина кратчайшей цепочки для заданного $n$ обозначается как $l(n) = r$

Цепочка называется звёздной, если для неё выполняется:

$$\forall i\in \left(0, r\right] : \exists k < i : a_i=a_{i-1}+a_k\quad $$

Справедливы следующие оценки для $l(n)$:

$$l(n) \leq \lambda(n) +\nu(n) - 1$$

$$l(n) \geq \lceil \log_{2}(n) \rceil$$

### Алгоритм Яо

Алгоритм Яо применяется для построения аддитивной цепочки для $n\in \mathbb{N}$.

Выбирается параметр $k > 1$, а $n$ раскладывается в системе счисления по основанию $2^{k}$:

$$n = \sum_{i = 0}^{j}a_i2^{ik}, \quad a_j \neq 0$$

Введём функцию $d(z) = \sum_{i:a_i = z}2^{ik}$
Представим шаги алгоритма:
<ol>
    <li>Составляется базоваю последовательность $\{1, 2, 4, \dots, 2^{\lambda(n)}\}$</li>
    <li>Вычисляется значение $d(z)$ для всех $z \in \{1, 2, 4, \dots,2^{k-1}\}, \quad d(z) \neq 0$</li>
    <li>Вычисляется $zd(z)$ для всех $z$ -- в итоге получается разложение $n = \sum_{z=1}^{2^k-1}zd(z)$</li>
    <li>Необходимые шаги добавляются в базовую последовательность</li>
</ol>

Алгоритм не гарантирует минимальности аддитивной цепочки.

### Алгоритм дробления вектора индексов
Вектором индексов называется последовательность $r_1,r_2,\dots,r_{m-1}$, где

$$r_i=\{z:1\leq z\leq i\}, \quad a_i=a_{i-1}+a_{r_{i-1}},\quad2\leq i\leq m-1.$$

Зададим два вектора индексов $r=\{r_i\}_{i=1}^{m-1}$ и $\tilde r=\{\tilde r_i\}_{i=1}^{m-1}$. Тогда $r \succ \tilde r$, если $r_1=\tilde r_1,r_2=\tilde r_2,\dots,r_{m-1}=\tilde r_{m-1}$, а $r_m>\tilde r_m$.

Рассмотрим вектор индекса вида $\{r_1, r_2, \dots, r_q\} \cup \{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}.$ Назовём левую часть фиксированной, а правую - меняющейся.

Значение последнего элемента цепочки наибольшее при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{q+1,\dots,m\}\quad a_{max}=a_{q+1}*2^{m-q}$

Значение последнего элемента цепочки наименьшее при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{1,\dots,1\}\quad a_{min}=a_{q+1}+m-q$  
Рассмотрим алгоритм дробления вектора индексов:
<ol>
    <li>Внешний цикл запускается по длинам $\underline l(n) \le m \le \overline l(n)$, выбирается индекс дроблений $q=m\div2$
    </li>
    <li>Внутренний цикл - перебор всех фиксированных частей. На каждом шагу вычисляются $a_{min}, a_{max}$ и строится цепочка.
        <ol>
          <li>Если $a_{m} = n$, то задача решена.</li>
          <li>Если $n \notin [a_{min}, a_{max}]$, то производится переход к следующему набору $\{r_1, r_2, \dots, r_q\}$</li>
          <li>Если $n \in [a_{min}, a_{max}]$, то производится перебора меняющейся части.
          <ol>
              <li>Если $a_{m} = a_{n}$, то задача решена</li>
              <li>Иначе производится переход к следующему (по введённой упорядоченности) разбиению на фиксированную и меняющуюся части $\{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}$</li>
         </ol>
         </li>
       </ol>
    </li>
    <li>Если все наборы фиксированной длины исчерпаны, то х длина инкрементируется во внешнем цикле.</li>
</ol>

Алгоритм гарантирует минимальность аддитивной цепочки.

### Гипотеза Шольца-Брауэра
Заключается в следующем:
$$l(2^{n}-1) \le l(n) + n - 1$$
Причём:
<ul>
    <li>Гипотеза доказана для звёздных цепочек при любых значениях $n$</li>
    <li>Гипотеза справедлива по меньшей мере для $n < 578469$</li>
    <li>Равенство выполняется при $n <= 64$</li>
</ul>

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минималь-
ных аддитивных цепочек с использованием системы компьютерной мате-
матики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы

### Алгоритм ЯО

Для реализации алгоритма Яо воспользуемся рядом вспомогатльных функций. Так, была реализована функция `digits(n, b)`, формирующая представление числа `n` в системе счисления по основанию `b`.

In [ ]:
def split_into_digits(number, base):
    if number == 0:
        return [0]
    digits = []
    while number > 0:
        digits += [number % base]
        number //= base
    return digits

Для вычисления значения $\mathrm{d}\!\left(z\right)$ была реализована функция `d(z, digits, k)`: 

In [2]:
def d(z, digits, k):
    return sum([2 ** (i * k) for i in range(len(digits)) if digits[i] == z])

В качестве вспомогательного был реализован метод SX аналогично прошлой работе:

In [3]:
def sx_sequence(n):
    if int(n) != n or n < 0:
        raise ValueError('Unexpected exponent encountered')
    if n == 0:
        return ''
    return str(bin(n))[3:].replace('0', 'S').replace('1','SX')

def sx_exponents(sx):
    y = 1
    steps = [y]
    for i, elem in enumerate(sx):
        y += 1 if elem == 'X' else y
        steps += [y]
    return steps

def lambda_(n):
    return len(str(bin(n))) - 3

В совокупности с предыдущими функциями представлен алгоритм Яо.

In [4]:
def yao_algoritm(n, k):
    base = 2 ** k
    yao_sequence = [2 ** i for i in range(lambda_(n) + 1)]
    digits = split_into_digits(n, base)
    ds = []  # values of d(z)
    for z in range(1, base):
        ds += [d(z, digits, k)]
        if ds[-1] != 0:
            for exponent in sx_exponents(sx_sequence(z)):
                yao_sequence.append(exponent * ds[z - 1])
    yao_sequence += [n]
    return sorted(set(yao_sequence))


Работа программы была проверена при всех $n \in \left\{14, 17, 534, 711, 1001, 1119, 1122, 1134, 2048\right\}$, $k \in \left\{2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12\right\}$:

In [5]:
def yao_test():
    ks = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
    ns = {
        14: [1, 2, 3, 6, 7, 14], 
        17: [1, 2, 4, 8, 16, 17],
        534: [1, 2, 4, 8, 16, 32, 33, 66, 132, 133, 266, 267, 534], 
        711: [1, 2, 4, 6, 7, 11, 22, 44, 88, 176, 352, 704, 711],  
        1001: [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001], 
        1119: [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119], 
        1122: [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122],  
        1134: [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134],
        2048: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
    }
    for n in ns:
        for k in ks:
            yao_result = yao_algoritm(n, k)
            yao_length = len(yao_result)
            shortest_result = ns[n]
            shortest_length = len(shortest_result)
            print(f"{k} | {n} | {yao_result} | {yao_length} | {shortest_result} | {shortest_length}")

yao_test()

2 | 14 | [1, 2, 4, 8, 12, 14] | 6 | [1, 2, 3, 6, 7, 14] | 6
3 | 14 | [1, 2, 3, 4, 6, 8, 14] | 7 | [1, 2, 3, 6, 7, 14] | 6
4 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
5 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
6 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
7 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
8 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
9 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
10 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
11 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
12 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6
2 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6
3 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6
4 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6
5 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6
6 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 

Результат был оформлен в виде таблицы:

 | k | n | Цепочка Яо | Длина цепочки Яо | Минимальная цепочка | Длина минимальной цепочки | 
 | - | - | ---------------------------- | ---------------------- | ------------------------- | ------------------- | 
| 2 | 14 | [1, 2, 4, 8, 12, 14] | 6 | [1, 2, 3, 6, 7, 14] | 6 |
| 3 | 14 | [1, 2, 3, 4, 6, 8, 14] | 7 | [1, 2, 3, 6, 7, 14] | 6 |
| 4 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 5 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 6 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 7 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 8 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 9 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 10 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 11 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 12 | 14 | [1, 2, 3, 4, 6, 7, 8, 14] | 8 | [1, 2, 3, 6, 7, 14] | 6 |
| 2 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 3 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 4 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 5 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 6 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 7 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 8 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 9 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 10 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 11 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 12 | 17 | [1, 2, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] | 6 |
| 2 | 1001 | [1, 2, 4, 8, 16, 20, 32, 40, 64, 128, 256, 320, 512, 640, 960, 1001] | 16 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 3 | 1001 | [1, 2, 4, 8, 16, 32, 40, 64, 128, 192, 256, 384, 448, 512, 513, 1001] | 16 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 4 | 1001 | [1, 2, 4, 8, 9, 16, 32, 48, 64, 96, 112, 128, 224, 256, 512, 768, 1001] | 17 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 5 | 1001 | [1, 2, 4, 8, 9, 16, 32, 64, 96, 128, 192, 224, 256, 448, 480, 512, 960, 992, 1001] | 19 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 6 | 1001 | [1, 2, 4, 5, 8, 10, 16, 20, 32, 40, 41, 64, 128, 192, 256, 384, 448, 512, 896, 960, 1001] | 21 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 7 | 1001 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 32, 52, 64, 104, 105, 128, 256, 384, 512, 768, 896, 1001] | 22 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 8 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 16, 28, 29, 32, 58, 64, 116, 128, 232, 233, 256, 512, 768, 1001] | 22 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 9 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 32, 60, 61, 64, 122, 128, 244, 256, 488, 489, 512, 1001] | 23 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 10 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 11 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 12 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 2 | 1119 | [1, 2, 4, 5, 8, 10, 15, 16, 32, 64, 128, 256, 512, 1024, 1104, 1119] | 16 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 3 | 1119 | [1, 2, 3, 4, 6, 7, 8, 16, 24, 32, 64, 128, 256, 512, 1024, 1119] | 16 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 4 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 32, 64, 80, 128, 256, 512, 1024, 1119] | 18 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 5 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 64, 128, 256, 512, 1024, 1119] | 19 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 6 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 64, 128, 256, 512, 1024, 1088, 1119] | 20 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 7 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 8 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 9 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 10 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 11 | 1119 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 68, 69, 128, 138, 139, 256, 278, 279, 512, 558, 559, 1024, 1118, 1119] | 23 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 12 | 1119 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 68, 69, 128, 138, 139, 256, 278, 279, 512, 558, 559, 1024, 1118, 1119] | 23 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 2 | 1001 | [1, 2, 4, 8, 16, 20, 32, 40, 64, 128, 256, 320, 512, 640, 960, 1001] | 16 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 3 | 1001 | [1, 2, 4, 8, 16, 32, 40, 64, 128, 192, 256, 384, 448, 512, 513, 1001] | 16 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 4 | 1001 | [1, 2, 4, 8, 9, 16, 32, 48, 64, 96, 112, 128, 224, 256, 512, 768, 1001] | 17 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 5 | 1001 | [1, 2, 4, 8, 9, 16, 32, 64, 96, 128, 192, 224, 256, 448, 480, 512, 960, 992, 1001] | 19 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 6 | 1001 | [1, 2, 4, 5, 8, 10, 16, 20, 32, 40, 41, 64, 128, 192, 256, 384, 448, 512, 896, 960, 1001] | 21 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 7 | 1001 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 32, 52, 64, 104, 105, 128, 256, 384, 512, 768, 896, 1001] | 22 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 8 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 16, 28, 29, 32, 58, 64, 116, 128, 232, 233, 256, 512, 768, 1001] | 22 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 9 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 32, 60, 61, 64, 122, 128, 244, 256, 488, 489, 512, 1001] | 23 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 10 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 11 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 12 | 1001 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 62, 64, 124, 125, 128, 250, 256, 500, 512, 1000, 1001] | 24 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 |
| 2 | 1119 | [1, 2, 4, 5, 8, 10, 15, 16, 32, 64, 128, 256, 512, 1024, 1104, 1119] | 16 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 3 | 1119 | [1, 2, 3, 4, 6, 7, 8, 16, 24, 32, 64, 128, 256, 512, 1024, 1119] | 16 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 4 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 32, 64, 80, 128, 256, 512, 1024, 1119] | 18 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 5 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 64, 128, 256, 512, 1024, 1119] | 19 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 6 | 1119 | [1, 2, 3, 4, 6, 7, 8, 14, 15, 16, 30, 31, 32, 64, 128, 256, 512, 1024, 1088, 1119] | 20 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 7 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 8 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 9 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 10 | 1119 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 47, 64, 94, 95, 128, 256, 512, 1024, 1119] | 21 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 11 | 1119 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 68, 69, 128, 138, 139, 256, 278, 279, 512, 558, 559, 1024, 1118, 1119] | 23 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 12 | 1119 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 68, 69, 128, 138, 139, 256, 278, 279, 512, 558, 559, 1024, 1118, 1119] | 23 | [1, 2, 3, 6, 9, 15, 30, 60, 69, 138, 276, 552, 1104, 1119] | 14 |
| 2 | 1122 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 128, 256, 512, 1024, 1088, 1122] | 15 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 3 | 1122 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1024, 1026, 1122] | 14 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 4 | 1122 | [1, 2, 4, 8, 16, 32, 48, 64, 96, 128, 256, 512, 1024, 1122] | 14 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 5 | 1122 | [1, 2, 4, 8, 16, 32, 64, 96, 128, 256, 512, 1024, 1122] | 13 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 6 | 1122 | [1, 2, 4, 8, 16, 17, 32, 34, 64, 128, 256, 512, 1024, 1088, 1122] | 15 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 7 | 1122 | [1, 2, 3, 4, 6, 8, 12, 16, 24, 32, 48, 49, 64, 98, 128, 256, 512, 1024, 1122] | 19 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 8 | 1122 | [1, 2, 3, 4, 6, 8, 12, 16, 24, 32, 48, 49, 64, 98, 128, 256, 512, 1024, 1122] | 19 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 9 | 1122 | [1, 2, 3, 4, 6, 8, 12, 16, 24, 32, 48, 49, 64, 98, 128, 256, 512, 1024, 1122] | 19 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 10 | 1122 | [1, 2, 3, 4, 6, 8, 12, 16, 24, 32, 48, 49, 64, 98, 128, 256, 512, 1024, 1122] | 19 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 11 | 1122 | [1, 2, 4, 8, 16, 17, 32, 34, 35, 64, 70, 128, 140, 256, 280, 512, 560, 561, 1024, 1122] | 20 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 12 | 1122 | [1, 2, 4, 8, 16, 17, 32, 34, 35, 64, 70, 128, 140, 256, 280, 512, 560, 561, 1024, 1122] | 20 | [1, 2, 4, 8, 16, 17, 34, 68, 136, 272, 544, 561, 1122] | 13 |
| 2 | 1134 | [1, 2, 4, 8, 12, 16, 17, 32, 34, 64, 128, 256, 512, 1024, 1088, 1134] | 16 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 3 | 1134 | [1, 2, 3, 4, 6, 8, 16, 32, 40, 64, 128, 256, 512, 1024, 1134] | 15 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 4 | 1134 | [1, 2, 3, 4, 6, 7, 8, 14, 16, 32, 48, 64, 96, 128, 256, 512, 1024, 1134] | 18 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 5 | 1134 | [1, 2, 3, 4, 6, 7, 8, 14, 16, 32, 64, 96, 128, 256, 512, 1024, 1134] | 17 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 6 | 1134 | [1, 2, 4, 5, 8, 10, 11, 16, 22, 23, 32, 46, 64, 128, 256, 512, 1024, 1088, 1134] | 19 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 7 | 1134 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 27, 32, 54, 55, 64, 110, 128, 256, 512, 1024, 1134] | 21 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 8 | 1134 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 27, 32, 54, 55, 64, 110, 128, 256, 512, 1024, 1134] | 21 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 9 | 1134 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 27, 32, 54, 55, 64, 110, 128, 256, 512, 1024, 1134] | 21 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 10 | 1134 | [1, 2, 3, 4, 6, 8, 12, 13, 16, 26, 27, 32, 54, 55, 64, 110, 128, 256, 512, 1024, 1134] | 21 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 11 | 1134 | [1, 2, 4, 8, 16, 17, 32, 34, 35, 64, 70, 128, 140, 141, 256, 282, 283, 512, 566, 567, 1024, 1134] | 22 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 12 | 1134 | [1, 2, 4, 8, 16, 17, 32, 34, 35, 64, 70, 128, 140, 141, 256, 282, 283, 512, 566, 567, 1024, 1134] | 22 | [1, 2, 3, 6, 7, 14, 28, 35, 70, 140, 280, 560, 567, 1134] | 14 |
| 2 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 3 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 4 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 5 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 6 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 7 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 8 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 9 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 10 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 11 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |
| 12 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 |

#### Вывод
Был реализован алгоритм Яо. Как видно из проведённых выше вычислений, длина цепочки Яо может как совпадать с длиной минимальной цепочки, так и значительно её превосходить, причём это зависит как от самого числа, так и от выбора параметра $k$. Тем не менее, преимуществом алгоритм Яо является высокая скорость его работы.

### Алгоритм дробления вектора индексов
Для $\mathrm{l}\!\left(n\right)$ справедливы следующие оценки:

$$\mathrm{l}\!\left(n\right) \geq \lambda\!\left(n\right) + \log_{2}\!\left(\nu\left(n\right)\right)$$

$$\mathrm{l}\!\left(n\right) \leq \lambda\!\left(n\right) + \nu\!\left(n\right) - 1$$

Они были реализованы следующим образом:

In [6]:
def lower_estimate(n):
    return lambda_(n) + int(log(list(bin(n)).count('1'), 2))

def upper_estimate(n):
    return lambda_(n) + list(bin(n)).count('1') - 1

Для построения звёздной цепочки по индексам была реализована функция `indices_to_sequence(indices)`:

In [7]:
def indices_to_sequence(indices):
    sequence = [1]
    for i in indices:
        sequence += [sequence[-1] + sequence[i - 1]]
    return sequence

Для осуществления перехода к следующему разбиению вектора индексов реализована функция `next_partition(partition, q = 0)`:

In [8]:
def next_partition(partition, q = 0):
    if partition == [1 for _ in range(len(partition))]:
        return []
    for i in range(len(partition) - 1, -1, -1):
        if partition[i] > 1:
            partition[i] -= 1
            break
        elif partition[i] == 1:
            partition[i] = i + q + 1
    return partition

Алгоритм дробления вектора индексов был реализован в функции `subdivision_algorithm(n)`:

In [9]:
def subdivision_algorithm(n):
    if n < 1:
        return []
    if n == 1:
        return [1]
    result = []
    for m in range(lower_estimate(n), upper_estimate(n) + 1):
        q = m // 2 + 1
        indices = [i for i in range(1, m + 1)]
        mutable_indices = indices[:q]
        constant_indices = indices[q:]
        while len(mutable_indices):
            sequence = indices_to_sequence(mutable_indices + constant_indices)
            a_min = sequence[q] + m - q
            a_max = sequence[q] * 2 ** (m - q)
            if sequence[-1] == n:
                return sequence
            elif n < a_min or n > a_max:
                mutable_indices = next_partition(mutable_indices)
            else:
                while len(constant_indices) :
                    sequence = indices_to_sequence(mutable_indices + constant_indices)
                    if n == sequence[-1]:
                        return sequence
                    constant_indices = next_partition(constant_indices, q)
                mutable_indices = next_partition(mutable_indices)
                constant_indices = indices[q:]

Корректность реализации была проверена для $n = 1001, 1119, 1122, 1134, 2048$.

In [10]:
import time

def subdivision_test():
    ns = [1001, 1119, 1122, 1134, 2048]
    
    for n in ns:
        start = time.time()
        sequence = subdivision_algorithm(n)
        duration = time.time() - start
        sequence_length = len(sequence)
        print(f' | {n} | {sequence} | {sequence_length} | {round(duration, 3)} секунд | ')

subdivision_test()

 | 1001 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001] | 14 | 703.657 секунд | 
 | 1119 | [1, 2, 4, 8, 16, 32, 64, 65, 81, 146, 292, 373, 746, 1119] | 14 | 557.148 секунд | 
 | 1122 | [1, 2, 4, 8, 16, 32, 64, 66, 132, 264, 528, 1056, 1122] | 13 | 3.464 секунд | 
 | 1134 | [1, 2, 4, 8, 16, 32, 64, 128, 160, 162, 324, 648, 972, 1134] | 14 | 478.253 секунд | 
 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | 0.0 секунд | 


Результаты проверки были оформлены в виде таблицы:

 |  n     |  Цепочка  |  Длина  |  Время работы  | 
 | ------ | --------------- | ----------------------------------------------------------------- | -------------- | 
 | 1001 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001] | 14 | 703.657 секунд | 
 | 1119 | [1, 2, 4, 8, 16, 32, 64, 65, 81, 146, 292, 373, 746, 1119] | 14 | 557.148 секунд | 
 | 1122 | [1, 2, 4, 8, 16, 32, 64, 66, 132, 264, 528, 1056, 1122] | 13 | 3.464 секунд | 
 | 1134 | [1, 2, 4, 8, 16, 32, 64, 128, 160, 162, 324, 648, 972, 1134] | 14 | 478.253 секунд | 
 | 2048 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048] | 12 | 0.0 секунд | 

#### Вывод
Был реализован алгоритм дробления вектора индексов. Как видно из проведённых выше результатов, он в среднем случае требует значительно больших затрат по времени по сравнению с алгоритмом Яо, но при этом гарантированно строит кратчайшую звёздную цепочку, что является его основным преимуществом.

### Гипотеза Шольца-Брауэра
Исходя из вышеизложенных теоритеческих положений, была реализована функция `modified_subdivision_algorithm(n, m)`, осуществляющая перебор только равных $\mathrm{l}\!\left(n\right)+n-1$ длин:

In [11]:
def modified_subdivision_algorithm(n, m):
    if n < 1:
        return []
    if n == 1:
        return [1]
    result = []
    m -= 1
    q = m // 2 + 1
    indices = [i for i in range(1, m + 1)]
    mutable_indices = indices[:q]
    constant_indices = indices[q:]
    while len(mutable_indices):
        sequence = indices_to_sequence(mutable_indices + constant_indices)
        a_min = sequence[q] + m - q
        a_max = sequence[q] * 2 ** (m - q)
        if sequence[-1] == n:
            return sequence
        elif n < a_min or n > a_max:
            mutable_indices = next_partition(mutable_indices)
        else:
            while len(constant_indices):
                sequence = indices_to_sequence(mutable_indices + constant_indices)
                if n == sequence[-1]:
                    return sequence
                constant_indices = next_partition(constant_indices, q)
            mutable_indices = next_partition(mutable_indices)
            constant_indices = indices[q:]

Для проверки гипотезы Шольца-Брауэра была реализована функция `sholz_brauer_test(left, right)`, проверяющая её для всех целочисленных значений $n$ в заданном отрезке:

In [12]:
def sholz_brauer_test(left, right):
    for i in range(left, right + 1):
        right_term = len(subdivision_algorithm(i)) + i - 1
        left_term = len(modified_subdivision_algorithm(2 ** i - 1, right_term))
        if left_term > right_term:
            print(f' | {i} | {left_term} | {right_term} | Неверна | ')
        else:
            print(f' | {i} | {left_term} | {right_term} | Верна | ')
        
sholz_brauer_test(1, 12)

 | 1 | 1 | 1 | Верна | 
 | 2 | 3 | 3 | Верна | 
 | 3 | 5 | 5 | Верна | 
 | 4 | 6 | 6 | Верна | 
 | 5 | 8 | 8 | Верна | 
 | 6 | 9 | 9 | Верна | 
 | 7 | 11 | 11 | Верна | 
 | 8 | 11 | 11 | Верна | 
 | 9 | 13 | 13 | Верна | 
 | 10 | 14 | 14 | Верна | 
 | 11 | 16 | 16 | Верна | 
 | 12 | 16 | 16 | Верна | 


Полученный вывод был оформлен в виде таблицы:

 | $n$ | $$\mathrm{l}\!\left(2^n-1\right)$$ | $$\mathrm{l}\!\left(n\right)+n-1$$ | Выполнение гипотезы
 | --- | --- | --- | --- | 
 | 1 | 1 | 1 | Верна | 
 | 2 | 3 | 3 | Верна | 
 | 3 | 5 | 5 | Верна | 
 | 4 | 6 | 6 | Верна | 
 | 5 | 8 | 8 | Верна | 
 | 6 | 9 | 9 | Верна | 
 | 7 | 11 | 11 | Верна | 
 | 8 | 11 | 11 | Верна | 
 | 9 | 13 | 13 | Верна | 
 | 10 | 14 | 14 | Верна | 
 | 11 | 16 | 16 | Верна | 
 | 12 | 16 | 16 | Верна | 

#### Вывод
Было проверена гипотеза Шольца-Брауэра. Установлено, что она корректна пр меньшей мере при $n \leq 12$. Для ускорения проверки был модифицирован алгоритм дробления вектора индексов посредством применения одного из следствий теоретических положений.

## Выводы

Были изучены аддитивные цепочки, а также их частный случай - звездные цепочки. Были освоены и на практике реализованы два алгоритма построения таких цепочек (алгоритм Яо и алгоритм дробления вектора индексов) с последующим сравнением, а также проверена гипотеза Шольца-Брауэра при $1 \leq n \leq 12$.

Было установлено несуществование совмещающего в себе одновременно низкую сложность по времени и гарантированное построение точного решения   алгоритма построения минимальной цепочки.

Были отточены навыки использования системы компьютерной математики SageMath для реализации алгоритмов.